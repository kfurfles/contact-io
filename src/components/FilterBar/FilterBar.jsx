import React, {useCallback, useEffect, useState} from 'react'
import Button from '../Button'
import Input from '../Input'

import './style.scss'

const initialState = {
    name: {
        selected: false,
        direction: 'down'
    },
    country: {
        selected: false,
        direction: 'down'
    },
    company: {
        selected: false,
        direction: 'down'
    },
    department: {
        selected: false,
        direction: 'down'
    },
    admissionDate: {
        selected: false,
        direction: 'down'
    }
}

const FilterBar = ({ onChangeFilter, onChangeSort }) => {

    const [ filterState, setFilterState ] = useState(initialState)
    const [filter, setFilter] = useState({})
    const [search, setSearch] = useState('');

    const {
        name,
        admissionDate,
        company,
        country,
        department
    } = filterState

    function setActiveFilter(filterValue, filterKey) {

        const { selected, direction } = filterState[filterKey]
        let newFilterKeyValue = {}

        if(selected){
            const newDirection = direction === 'up' ? 'down' : 'up'
            newFilterKeyValue = {
                direction: newDirection,
                selected: true,
            }
        } else {
            newFilterKeyValue = {
                direction: 'up',
                selected: true,
            }
        }

        setFilterState({
            ...initialState,
            [filterKey]: {
                ...newFilterKeyValue
            }
        })
        setFilter({ ...newFilterKeyValue, key: filterKey})
        setSearch('')
        onChangeSort({ key: filterKey, direction: newFilterKeyValue.direction })
    }

    const setSearchFilter = useCallback(e =>{
        const value = e.target.value
        setSearch(value)
        onChangeFilter({ value, ...filter })
        // eslint-disable-next-line react-hooks/exhaustive-deps
    },[filter])

    useEffect(() =>{
        setActiveFilter(initialState.name, 'name')
        // eslint-disable-next-line react-hooks/exhaustive-deps
    },[])

    return (
        <section className="filters">
        <div className="filters__search">
            <Input
                value={search}
                onChange={setSearchFilter}
                className="filters__search__input"
                type="text"  placeholder="Pesquisar" />

            <button className="filters__search__icon">
                <i className="fa fa-search"/>
            </button>
        </div>

        <Button {...name} onClick={() => setActiveFilter(name,'name')}>Nome</Button>

        <Button {...country} onClick={() => setActiveFilter(name,'country')}>País</Button>
        <Button {...company} onClick={() => setActiveFilter(name,'company')}>Empresa</Button>
        <Button {...department} onClick={() => setActiveFilter(name,'department')}>Departamento</Button>
        <Button {...admissionDate} onClick={() => setActiveFilter(name,'admissionDate')}>Data de admissão</Button>
      </section>
    )
}

export default FilterBar