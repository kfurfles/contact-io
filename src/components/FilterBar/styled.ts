import styled from '@emotion/styled'
import Input from '../Input'

export const ButtonBase = styled.button`
    background-color: transparent;
    border: none;
    box-shadow: none;
    outline: none;
`

export const Container = styled.section`
    align-items: center;
    display: flex;
    flex-wrap: nowrap;
    padding: 24px 0;
`

export const SearchBox = styled.div`
    background: #fff;
    border-radius: 5px;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 2px 2px rgba(0, 0, 0, .07);
    display: flex;
    flex-wrap: nowrap;
    height: 40px;
    justify-content: stretch;
    margin-right: 8px;
    overflow: hidden;
    width: 100%;
`

export const ButtonFilter = styled(ButtonBase)`
    background: transparent;
    font-size: 1.4rem;
    color: #9FA9BC;

    &:hover{
        color: #4C84FF;
    }
`

export const InputFilter = styled(Input)`
    background: transparent;
    border: none;
    color: #4C84FF;
    font-size: 1.4rem;
    font-weight: bold;
    outline: none;
    padding: 5px 10px;
    width: 100%;
    
    &::placeholder {
        color: #9FA9BC;
        font-weight: normal;
    }
`
