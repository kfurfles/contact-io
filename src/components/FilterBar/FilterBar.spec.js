import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';

import FilterBar from './index';

const mockUserData = {
	avatar: '',
	name: '',
	phone: '',
	country: '',
	admissionDate: '',
	company: '',
	department: '',
};

describe('FilterBar', () => {
	it('Should render FilterBar', () => {
		const { fragment } = render(<FilterBar onChangeSort={() => {}} onChangeFilter={() => {}} />);

		expect(fragment).toMatchSnapshot();
	});

	it('Should dispatch {setActiveFilter}', () => {
		const handleChangeSort = (a) => {
			expect()
		}
		const handleChangeFilter = (e) => {
			console.log({ e })
		}
		const filterBar = render(<FilterBar onChangeSort={handleChangeSort} onChangeFilter={handleChangeFilter} />);
		filterBar.getByText('Nome').click()
	})
});