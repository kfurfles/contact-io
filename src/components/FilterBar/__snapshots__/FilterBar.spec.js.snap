// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar Should render FilterBar 1`] = `undefined`;
